staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go

```diff
@@ -1407,3 +1409,193 @@ func isStreamReset(err error) bool {
+					parentGot = parent
+					return &fakeContext{deadline: at}, func() {}
+				},
+			}
+
+			queueWaitCtxGot, _ := qw.GetRequestWaitContext(parent, test.isWatch)
+			switch {
+			case test.reqWaitCtxExpected:
+				deadlineGot, _ := queueWaitCtxGot.Deadline()
+				if got := deadlineGot.Sub(now); got != test.reqWaitLimitExpected {
+					t.Errorf("Expected queue wait time %s, but got: %s", test.reqWaitLimitExpected, got)
+				}
+				if parent != parentGot {
+					t.Errorf("Expected the parent to be used when creating the child context")
+				}
+			default:
+				if queueWaitCtxGot != parent {
+					t.Errorf("Expected the parent context to be returned: want: %#v, got %#v", parent, queueWaitCtxGot)
+				}
+				if parentGot != nil {
+					t.Errorf("Did not expect the new context func to be invoked")
+				}
+			}
+		})
+	}
+}
+
+func TestGetRequestWaitContext(t *testing.T) {
+	tests := []struct {
+		name                    string
+		defaultRequestWaitLimit time.Duration
+		isWatch                 bool
+		parent                  func(t time.Time) (context.Context, context.CancelFunc)
+		reqWaitLimitExpected    time.Duration
+	}{
+		{
+			name: "parent context has a deadline, queue wait deadline should be approximately one fourth",
+			parent: func(t time.Time) (context.Context, context.CancelFunc) {
+				return context.WithDeadline(context.Background(), t.Add(80*time.Second))
+			},
+			reqWaitLimitExpected: 20 * time.Second,
+		},
+		{
+			name:                    "parent does not have any deadline",
+			defaultRequestWaitLimit: 15 * time.Second,
+			parent: func(_ time.Time) (context.Context, context.CancelFunc) {
+				return context.WithCancel(context.Background())
+			},
+			reqWaitLimitExpected: 15 * time.Second,
+		},
+		{
+			name:                    "request is a watch",
+			isWatch:                 true,
+			defaultRequestWaitLimit: 15 * time.Second,
+			parent: func(t time.Time) (context.Context, context.CancelFunc) {
+				return context.WithCancel(context.Background())
+			},
+			reqWaitLimitExpected: 15 * time.Second,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			now := time.Now()
+			parent, cancel := test.parent(now)
+			defer cancel()
+
+			clock := clocktesting.NewFakePassiveClock(now)
+			qw := newRequestWaitContextFactory(clock, test.defaultRequestWaitLimit)
+			queueWaitCtxGot, cancelGot := qw.GetRequestWaitContext(parent, test.isWatch)
+			if cancelGot == nil {
+				t.Errorf("Expected a non nil context.CancelFunc")
+				return
+			}
+			defer cancelGot()
+
+			deadlineGot, ok := queueWaitCtxGot.Deadline()
+			if !ok {
+				t.Errorf("Expected the context to have a deadline")
+			}
+			// we can't do an exact comparison here, let's keep some gap to ensure it does
+			// not flake in CI, we already have exact checks with the fake context tests.
+			if got := time.Until(deadlineGot); got < test.reqWaitLimitExpected/2 {
```

Comment:
The thing being tested is using the fake clock, why is this statement using the real clock?  Why not `got := deadlineGot.Sub(clock.Now())`?